tests/unit/models/gpflow/test_models.py

```diff
@@ -1626,3 +1628,331 @@ def test_sparse_variational_pairwise_covariance_for_non_whitened(
+
+    search_space = Box([0.0], [10.0])
+
+    model = MultifidelityAutoregressive(
+        build_multifidelity_autoregressive_models(
+            dataset, num_fidelities=2, input_search_space=search_space
+        )
+    )
+
+    starting_f0_model_loss = model.lowest_fidelity_signal_model.model.training_loss()
+    starting_f1_model_loss = model.fidelity_residual_models[1].model.training_loss()
+
+    model.update(dataset)
+    model.optimize(dataset)
+
+    assert model.lowest_fidelity_signal_model.model.training_loss() < starting_f0_model_loss
+    assert model.fidelity_residual_models[1].model.training_loss() < starting_f1_model_loss
+
+
+@pytest.mark.parametrize(
+    "new_data,problem",
+    (
+        ([[0.0, 8.0]], "too_high_fid"),
+        ([[0.0, -1.0]], "negative_fid"),
+        ([[0.0, 1.3]], "non_int_fid"),
+    ),
+)
+def test_multifidelity_autoregressive_optimize_raises_for_bad_new_data(
+    new_data: list[list[float]], problem: str
+) -> None:
+
+    new_dataset = Dataset(
+        tf.Variable(new_data, dtype=tf.float64), tf.Variable([[0.1]], dtype=tf.float64)
+    )
+    model = multifidelity_autoregressive_model()
+    dataset = multifidelity_autoregressive_nd_dataset()
+    if problem == "non_int_fid":
+        with pytest.raises(tf.errors.InvalidArgumentError):
+            model.optimize(dataset + new_dataset)
+    else:
+        with pytest.raises(ValueError):
+            model.optimize(dataset + new_dataset)
+
+
+def test_multifidelity_autoregressive_sample_aligns_with_predict() -> None:
+
+    xs_low = tf.Variable(np.linspace(0, 10, 100), dtype=tf.float64)[:, None]
+    xs_high = tf.Variable(np.linspace(0, 10, 10), dtype=tf.float64)[:, None]
+    lf_obs = tf.sin(xs_low)
+    hf_obs = 2 * tf.sin(xs_high) + tf.random.normal(
+        xs_high.shape, mean=0, stddev=1e-1, dtype=tf.float64
+    )
+
+    lf_query_points = add_fidelity_column(xs_low, 0)
+    hf_query_points = add_fidelity_column(xs_high, 1)
+
+    lf_dataset = Dataset(lf_query_points, lf_obs)
+    hf_dataset = Dataset(hf_query_points, hf_obs)
+
+    dataset = lf_dataset + hf_dataset
+
+    search_space = Box([0.0], [10.0])
+
+    model = MultifidelityAutoregressive(
+        build_multifidelity_autoregressive_models(
+            dataset, num_fidelities=2, input_search_space=search_space
+        )
+    )
+
+    model.lowest_fidelity_signal_model.model.likelihood.variance.assign(1.1e-6)
+    gpflow.set_trainable(model.lowest_fidelity_signal_model.model.likelihood, False)
+
+    model.update(dataset)
+    model.optimize(dataset)
+
+    test_locations = tf.Variable(np.linspace(0, 10, 32), dtype=tf.float64)[:, None]
+    lf_test_locations = add_fidelity_column(test_locations, 0)
+    hf_test_locations = add_fidelity_column(test_locations, 1)
+
+    lf_true_means, lf_true_vars = model.predict(lf_test_locations)
+
+    hf_true_means, hf_true_vars = model.predict(hf_test_locations)
+
+    lf_samples = model.sample(lf_test_locations, 100000)
```

Comment:
maybe best to merge the two `sample` calls into one? to also capture a possible bug if we get it wrong with multiple fidelity calls at the same time